src/v/archival/ntp_archiver_service.cc

```diff
@@ -612,59 +659,99 @@ ntp_archiver::schedule_uploads(model::offset last_stable_offset) {
-                           return ss::make_ready_future<ss::stop_iteration>(
-                             res);
-                       });
-                 })
-            .then([&scheduled] {
-                return ss::make_ready_future<std::vector<scheduled_upload>>(
-                  std::move(scheduled));
-            });
-      });
+      last_stable_offset,
+      allow_reuploads_t::no);
+
+    if (
+      config::shard_local_cfg().cloud_storage_enable_compacted_topic_reupload()
+      && _partition->get_ntp_config().is_compacted()) {
+        params.emplace_back(
+          segment_upload_kind::compacted,
+          compacted_segments_upload_start,
+          model::offset::max(),
+          allow_reuploads_t::yes);
+    }
+
+    co_return co_await schedule_uploads(std::move(params));
 }
 
-ss::future<ntp_archiver::batch_result> ntp_archiver::wait_all_scheduled_uploads(
-  std::vector<ntp_archiver::scheduled_upload> scheduled) {
-    ntp_archiver::batch_result total{};
+ss::future<std::vector<ntp_archiver::scheduled_upload>>
+ntp_archiver::schedule_uploads(std::vector<upload_context> loop_contexts) {
+    std::vector<scheduled_upload> scheduled_uploads;
+    auto uploads_remaining = _concurrency;
+    for (auto& ctx : loop_contexts) {
+        if (uploads_remaining <= 0) {
+            vlog(
+              _rtclog.info,
+              "no more upload slots remaining, skipping upload kind: {}, start "
+              "offset: {}, last offset: {}, uploads remaining: {}",
+              ctx.upload_kind,
+              ctx.start_offset,
+              ctx.last_offset,
+              uploads_remaining);
+            break;
+        }
+
+        vlog(
+          _rtclog.debug,
+          "scheduling uploads, start offset: {}, last offset: {}, upload kind: "
+          "{}, uploads remaining: {}",
+          ctx.start_offset,
+          ctx.last_offset,
+          ctx.upload_kind,
+          uploads_remaining);
+
+        // this metric is only relevant for non compacted uploads.
+        if (ctx.upload_kind == segment_upload_kind::non_compacted) {
+            _probe.upload_lag(ctx.last_offset - ctx.start_offset);
+        }
+
+        co_await ss::repeat(
+          [this, &ctx, &uploads_remaining]() -> ss::future<ss::stop_iteration> {
+              if (uploads_remaining <= 0 || !upload_loop_can_continue()) {
+                  co_return ss::stop_iteration::yes;
+              }
+
+              auto should_stop = co_await ctx.schedule_single_upload(*this);
+
+              // At the end of each context is an upload with should_stop=yes,
+              // which does not upload anything but signals the end of the
+              // uploads. Do not decrement remaining uploads for this type of
+              // return value. When should_stop=no, we have a real upload and
+              // decrement the counter.
+              if (should_stop == ss::stop_iteration::no) {
+                  uploads_remaining -= 1;
```

Comment:
you cannot access captures after co_await in a lambda

Reply:
I am cleaning these up in #7053